packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -39,195 +45,187 @@ export const DECODE_CONTRACT_CALL_RESULT_ABI = [
     ) => Vow<any>; // XXX HostForGuest???
+    timer: Remote<TimerService>;
     proposalShapes: ReturnType<typeof makeProposalShapes>;
+    vowTools: VowTools;
+    zcf: ZCF;
   },
 ) =>
   zone.exoClassKit(
     'Portfolio',
     {
+      tap: M.interface('tap', {
+        receiveUpcall: M.call(M.record()).returns(M.promise()),
+      }),
       keeper: KeeperI,
       rebalanceHandler: OfferHandlerI,
       invitationMakers: M.interface('invitationMakers', {
         Rebalance: M.callWhen().returns(InvitationShape),
       }),
-      tap: EvmTapI,
     },
-
-    (localAccount: LocalAccount): PortfolioKitState => ({
-      localAccount,
-      Aave: undefined,
-      Compound: undefined,
-      USDN: undefined,
-    }),
+    (): PortfolioKitState => {
+      return {
+        Aave: undefined,
+        Compound: undefined,
+        USDN: undefined,
+        Gmp: undefined,
+      };
+    },
     {
       tap: {
         async receiveUpcall(event: VTransferIBCEvent) {
           trace('receiveUpcall', event);
 
-          const tx: FungibleTokenPacketData = JSON.parse(
-            atob(event.packet.data),
-          );
+          // TODO: dont use the same LCA for sending transfers to noble
+          let tx: FungibleTokenPacketData;
+          try {
+            tx = JSON.parse(atob(event.packet.data));
+          } catch (err) {
+            trace(
+              'Failed to parse packet data JSON in receiveUpcall:',
+              err,
+              event.packet.data,
+            );
+            return;
+          }
 
           trace('receiveUpcall packet data', tx);
           if (!tx.memo) return;
-          const memo: AxelarGmpIncomingMemo = JSON.parse(tx.memo); // XXX unsound! use typed pattern
-
-          if (!(memo.source_chain in PositionChain)) {
-            console.warn('unknown source_chain', memo);
+          let memo: AxelarGmpIncomingMemo;
+          try {
+            memo = JSON.parse(tx.memo); // XXX unsound! use typed pattern
+          } catch (err) {
+            trace('Failed to parse memo JSON in receiveUpcall:', err, tx.memo);
             return;
           }
 
-          // const payloadBytes = decodeBase64(memo.payload.replace(/^0x/, ''));
+          const ids = values(axelarChainsMap).map(chain => chain.axelarId);
+          if (!ids.includes(memo.source_chain)) return;
+
+          const payloadBytes = decodeBase64(memo.payload);
           const [{ isContractCallResult, data }] = decodeAbiParameters(
             DECODE_CONTRACT_CALL_RESULT_ABI,
-            memo.payload as `0x${string}`, // hm.. cast...
+            payloadBytes,
           ) as [AgoricResponse];
 
-          trace(
-            'receiveUpcall Decoded:',
-            JSON.stringify({ isContractCallResult, data }),
-          );
-
-          if (isContractCallResult) {
-            console.warn('TODO: Handle the result of the contract call', data);
-          } else {
+          if (!isContractCallResult) {
```

Comment:
`if (!x) { ... } else { ... }` is odd, to my eye. I **suggest**:

```suggestion
          if (isContractCallResult) {
            // TODO Handle the result of the contract call
          } else {
```